snuba/query/allocation_policies/per_referrer.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+import logging
+
+from snuba.query.allocation_policies import (
+    AllocationPolicyConfig,
+    QueryResultOrError,
+    QuotaAllowance,
+)
+from snuba.query.allocation_policies.concurrent_rate_limit import (
+    BaseConcurrentRateLimitAllocationPolicy,
+)
+from snuba.redis import RedisClientKey, get_redis_client
+from snuba.state.rate_limit import RateLimitParameters
+
+rds = get_redis_client(RedisClientKey.RATE_LIMITER)
+
+logger = logging.getLogger("snuba.query.allocation_policy_per_referrer")
+
+_DEFAULT_MAX_THREADS = 10
+
+
+class ReferrerGuardRailPolicy(BaseConcurrentRateLimitAllocationPolicy):
+    """
+    A policy to prevent runaway referrers from consuming too many queries. This concern is orthogonal to customer rate limits.
```

Comment:
Something to address here: customers can absolutely trip this limit, if a single customer decides to abuse a specific endpoint.

So on the surface this doesn't seem orthogonal to customers at all. We have a separate policy to handle customer+referrer limits, but from this comment it isn't clear when to use one over the other.